lib/chef/provisioning/aws_driver/driver.rb

```diff
@@ -102,22 +117,111 @@ def allocate_load_balancer(action_handler, lb_spec, lb_options, machine_specs)
+          if current != desired
+            perform_action.call("  updating security groups to #{desired.to_a}") do
+              elb.client.apply_security_groups_to_load_balancer(
+                load_balancer_name: actual_elb.name,
+                security_groups: desired.to_a
+              )
+            end
           end
         end
-        if disable_zones.size > 0
-          perform_action.call("  disable availability zones #{disable_zones.join(', ')}") do
-            actual_elb.availability_zones.disable(*disable_zones)
+
+        # A subnet always belongs to an availability zone.  When specifying a ELB spec, you can either
+        # specify subnets OR AZs but not both.  You cannot specify multiple subnets in the same AZ.
+        # You must specify at least 1 subnet or AZ.  On an update you cannot remove all subnets
+        # or AZs - it must belong to one.
+        if !availability_zones.empty? && !subnets.empty?
+          # We do this check here because there is no atomic call we can make to specify both
+          # subnets and AZs at the same time
+          raise "You cannot specify both `availability_zones` and `subnets`"
+        else
+          # Users can switch from availability zones to subnets or vice versa.  To ensure we do not
+          # unassign all (which causes an AWS error) we first add all available ones, then remove
+          # an unecessary ones
+          actual_zones_subnets = {}
+          actual_elb.subnets.each do |subnet|
+            actual_zones_subnets[subnet.id] = subnet.availability_zone.name
+          end
+
+          # Only 1 of subnet or AZ will be populated b/c of our check earlier
+          desired_subnets_zones = {}
+          availability_zones.each do |zone|
+            # If the user specifies availability zone, we find the default subnet for that
+            # AZ because this duplicates the create logic
+            zone = zone.downcase
+            filters = [
+              {:name => 'availabilityZone', :values => [zone]},
+              {:name => 'defaultForAz', :values => ['true']}
+            ]
+            default_subnet = ec2.client.describe_subnets(:filters => filters)[:subnet_set]
+            if default_subnet.size != 1
+              raise "Could not find default subnet in availability zone #{zone}"
+            end
+            default_subnet = default_subnet[0]
+            desired_subnets_zones[default_subnet[:subnet_id]] = zone
+          end
+          unless subnets.empty?
+            subnet_query = ec2.client.describe_subnets(:subnet_ids => subnets)[:subnet_set]
+            # AWS raises an error on an unknown subnet, but not an unknown AZ
+            subnet_query.each do |subnet|
+              zone = subnet[:availability_zone].downcase
+              desired_subnets_zones[subnet[:subnet_id]] = zone
+            end
+          end
+
+          # We only bother attaching subnets, because doing this automatically attaches the AZ
```

Comment:
Oh wow!  So what you are saying is, when you attach the subnet AZ-1-defaultsubnet, and then describe the LB, amazon reports it as attached to the AZ itself rather than the subnet?